kythe/go/extractors/config/runextractor/mavencmd/pom_xml_modifier.go

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2018 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mavencmd
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/beevik/etree"
+)
+
+// PreProcessPomXML takes a pom.xml file and either verifies that it already has
+// the bits necessary to specify a separate compiler on commandline, or adds
+// functionality by dropping in a maven-compiler-plugin to the build.
+//
+// Note this potentially modifies the input file, so make a copy beforehand if
+// you need to keep the original.
+func PreProcessPomXML(pomXMLFile string) error {
+	doc := etree.NewDocument()
+	err := doc.ReadFromFile(pomXMLFile)
+	if err != nil {
+		return fmt.Errorf("reading XML file %s: %v", pomXMLFile, err)
+	}
+	if hasCompilerPlugin(doc) {
+		return nil
+	}
+	if err := appendCompilerPlugin(doc); err != nil {
+		return err
+	}
+	f, err := os.OpenFile(pomXMLFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("opening file %s for append: %v", pomXMLFile, err)
+	}
+	doc.Indent(2)
+	doc.WriteTo(f)
+	return f.Close()
+}
+
+func hasCompilerPlugin(doc *etree.Document) bool {
+	for _, p := range doc.FindElements("//project/build/plugins/plugin/artifactId") {
+		if p.Text() == "maven-compiler-plugin" {
+			return true
+		}
+	}
+	return false
+}
+
+func appendCompilerPlugin(doc *etree.Document) error {
```

Comment:
Please add a comment describing roughly what this does, something like

```
// appendCompilerPlugin modifies a doc containing the root element of a pom.xml
// configuration, by injecting a custom Maven compiler <plugin> element into the 
// plugin list for the top-level project.
```

or whatever is actually true. Sometimes a schematic example can help, YMMV.

Reply:
Oo I wonder if pointing to the test files is a good sufficient example

Reply:
Also including a comment to make clear the difference between maven & gradle, since it's a bit funky to do the preprocess for maven and then also still have to pass compiler flags